{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "mount_file_id": "1vqrkkT3_JrxjFeV1vuOwSznqR-iVcDDf",
      "authorship_tag": "ABX9TyNcRark02g7bixejqAx8+kr",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/eoinleen/Protein-design-random/blob/main/20250129-dickin-about-Seq_analysis_RFdiff_v10.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "AJ2WyL3YufKK",
        "outputId": "82a8ff2f-0e1c-4260-d78c-587f18345ed1"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: biopython in /usr/local/lib/python3.11/dist-packages (1.85)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from biopython) (1.26.4)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (1.26.4)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.11/dist-packages (2.2.2)\n",
            "Requirement already satisfied: numpy>=1.23.2 in /usr/local/lib/python3.11/dist-packages (from pandas) (1.26.4)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas) (2024.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas) (2025.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas) (1.17.0)\n",
            "Requirement already satisfied: seaborn in /usr/local/lib/python3.11/dist-packages (0.13.2)\n",
            "Requirement already satisfied: numpy!=1.24.0,>=1.20 in /usr/local/lib/python3.11/dist-packages (from seaborn) (1.26.4)\n",
            "Requirement already satisfied: pandas>=1.2 in /usr/local/lib/python3.11/dist-packages (from seaborn) (2.2.2)\n",
            "Requirement already satisfied: matplotlib!=3.6.1,>=3.4 in /usr/local/lib/python3.11/dist-packages (from seaborn) (3.10.0)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (1.3.1)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.11/dist-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (4.55.6)\n",
            "Requirement already satisfied: kiwisolver>=1.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (1.4.8)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (24.2)\n",
            "Requirement already satisfied: pillow>=8 in /usr/local/lib/python3.11/dist-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (11.1.0)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (3.2.1)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.11/dist-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas>=1.2->seaborn) (2024.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas>=1.2->seaborn) (2025.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.7->matplotlib!=3.6.1,>=3.4->seaborn) (1.17.0)\n",
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.11/dist-packages (3.10.0)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (1.3.1)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (4.55.6)\n",
            "Requirement already satisfied: kiwisolver>=1.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (1.4.8)\n",
            "Requirement already satisfied: numpy>=1.23 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (1.26.4)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (24.2)\n",
            "Requirement already satisfied: pillow>=8 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (11.1.0)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (3.2.1)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (2.8.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.7->matplotlib) (1.17.0)\n",
            "Requirement already satisfied: scikit-learn in /usr/local/lib/python3.11/dist-packages (1.6.1)\n",
            "Requirement already satisfied: numpy>=1.19.5 in /usr/local/lib/python3.11/dist-packages (from scikit-learn) (1.26.4)\n",
            "Requirement already satisfied: scipy>=1.6.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn) (1.13.1)\n",
            "Requirement already satisfied: joblib>=1.2.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn) (1.4.2)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn) (3.5.0)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.11/dist-packages (1.13.1)\n",
            "Requirement already satisfied: numpy<2.3,>=1.22.4 in /usr/local/lib/python3.11/dist-packages (from scipy) (1.26.4)\n",
            "Requirement already satisfied: joblib in /usr/local/lib/python3.11/dist-packages (1.4.2)\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "FileNotFoundError",
          "evalue": "Input file not found: /content/drive/MyDrive/Fasta-files/3NOB_90-110/3NOB_90-110_design.fasta",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-2-3838f00959aa>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    178\u001b[0m     \u001b[0minput_file\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"/content/drive/MyDrive/Fasta-files/3NOB_90-110/3NOB_90-110_design.fasta\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    179\u001b[0m     \u001b[0moutput_dir\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"/content/drive/MyDrive/Fasta-files/3NOB_90-110/analysis_output\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 180\u001b[0;31m     \u001b[0manalyze_sequences\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput_dir\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;32m<ipython-input-2-3838f00959aa>\u001b[0m in \u001b[0;36manalyze_sequences\u001b[0;34m(input_file, output_dir)\u001b[0m\n\u001b[1;32m    152\u001b[0m     \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmakedirs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput_dir\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexist_ok\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    153\u001b[0m     \u001b[0mextracted_file\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput_dir\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"extracted_sequences.fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 154\u001b[0;31m     \u001b[0msequences\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mextract_designed_sequences\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mextracted_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    155\u001b[0m     \u001b[0msequence_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mseq\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msequences\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    156\u001b[0m     \u001b[0msequence_names\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mseq\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msequences\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m<ipython-input-2-3838f00959aa>\u001b[0m in \u001b[0;36mextract_designed_sequences\u001b[0;34m(input_file, output_file)\u001b[0m\n\u001b[1;32m     52\u001b[0m     \u001b[0;34m\"\"\"Extract designed sequences from RF_diffusion output and save to a new FASTA file.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     53\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 54\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mFileNotFoundError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"Input file not found: {input_file}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     55\u001b[0m     \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmakedirs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdirname\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexist_ok\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     56\u001b[0m     \u001b[0mextracted_sequences\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFileNotFoundError\u001b[0m: Input file not found: /content/drive/MyDrive/Fasta-files/3NOB_90-110/3NOB_90-110_design.fasta"
          ]
        }
      ],
      "source": [
        "\"\"\"\n",
        "Script: Sequence Analysis Pipeline with Cluster Images\n",
        "Author: [Your Name]\n",
        "Date: [Current Date]\n",
        "\n",
        "Description:\n",
        "This script performs a sequence analysis pipeline, which includes the following steps:\n",
        "1. Extracting designed sequences from an input FASTA file.\n",
        "2. One-hot encoding the extracted sequences.\n",
        "3. Computing pairwise Hamming distances between the encoded sequences.\n",
        "4. Creating and saving a phylogenetic tree visualization based on the distance matrix.\n",
        "5. Generating a similarity heatmap for the sequences.\n",
        "6. Exporting individual images for each cluster in the dendrogram.\n",
        "\n",
        "Dependencies:\n",
        "- Biopython\n",
        "- NumPy\n",
        "- Pandas\n",
        "- Seaborn\n",
        "- Matplotlib\n",
        "- Scikit-learn\n",
        "- SciPy\n",
        "- Joblib\n",
        "- Regular Expressions (re)\n",
        "\"\"\"\n",
        "\n",
        "# Install required packages\n",
        "!pip install biopython\n",
        "!pip install numpy\n",
        "!pip install pandas\n",
        "!pip install seaborn\n",
        "!pip install matplotlib\n",
        "!pip install scikit-learn\n",
        "!pip install scipy\n",
        "!pip install joblib\n",
        "\n",
        "import os\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "from Bio import SeqIO\n",
        "from Bio.Seq import Seq\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "from sklearn.cluster import AgglomerativeClustering\n",
        "from scipy.spatial.distance import pdist, squareform\n",
        "from scipy.cluster import hierarchy\n",
        "from joblib import Parallel, delayed\n",
        "import re\n",
        "\n",
        "def extract_designed_sequences(input_file, output_file):\n",
        "    \"\"\"Extract designed sequences from RF_diffusion output and save to a new FASTA file.\"\"\"\n",
        "    if not os.path.exists(input_file):\n",
        "        raise FileNotFoundError(f\"Input file not found: {input_file}\")\n",
        "    os.makedirs(os.path.dirname(output_file), exist_ok=True)\n",
        "    extracted_sequences = []\n",
        "    design_lengths = set()\n",
        "\n",
        "    for idx, record in enumerate(SeqIO.parse(input_file, \"fasta\")):\n",
        "        try:\n",
        "            match = re.search(r'design:(\\d+).*n:(\\d+)', record.description)\n",
        "            if match:\n",
        "                new_id = f\"d{match.group(1)}_n{match.group(2)}\"\n",
        "\n",
        "                if '/' in str(record.seq):\n",
        "                    designed_seq = str(record.seq).split('/')[1].replace('-', '').strip()\n",
        "                    design_lengths.add(len(designed_seq))\n",
        "                    extracted_sequences.append(SeqRecord(Seq(designed_seq), id=new_id, description=\"\"))\n",
        "                else:\n",
        "                    print(f\"Warning: No '/' found in sequence {record.id}\")\n",
        "            else:\n",
        "                print(f\"Warning: Could not parse sequence {idx}\")\n",
        "        except Exception as e:\n",
        "            print(f\"Error processing sequence {idx}: {e}\")\n",
        "\n",
        "    if not extracted_sequences:\n",
        "        raise ValueError(\"No valid sequences found in input file\")\n",
        "\n",
        "    SeqIO.write(extracted_sequences, output_file, \"fasta\")\n",
        "    print(f\"Extracted {len(extracted_sequences)} sequences, saved to: {output_file}\")\n",
        "    return extracted_sequences\n",
        "\n",
        "def one_hot_encode(sequences):\n",
        "    \"\"\"One-hot encode amino acid sequences.\"\"\"\n",
        "    amino_acids = \"ACDEFGHIKLMNPQRSTVWY\"\n",
        "    aa_dict = {aa: i for i, aa in enumerate(amino_acids)}\n",
        "    encoding = np.zeros((len(sequences), len(sequences[0]), len(amino_acids)))\n",
        "\n",
        "    for i, seq in enumerate(sequences):\n",
        "        for j, char in enumerate(seq):\n",
        "            if char in aa_dict:\n",
        "                encoding[i, j, aa_dict[char]] = 1\n",
        "    return encoding.reshape(len(sequences), -1)\n",
        "\n",
        "def compute_distance_matrix(encoded_seqs):\n",
        "    \"\"\"Compute pairwise Hamming distances using parallel processing.\"\"\"\n",
        "    return squareform(pdist(encoded_seqs, metric=\"hamming\"))\n",
        "\n",
        "def create_phylogenetic_tree(distance_matrix, sequence_names, output_dir, num_clusters=3):\n",
        "    \"\"\"Create and save a phylogenetic tree visualization, and separate cluster images.\"\"\"\n",
        "    os.makedirs(output_dir, exist_ok=True)\n",
        "\n",
        "    if distance_matrix.shape[0] != len(sequence_names):\n",
        "        raise ValueError(f\"Mismatch: distance matrix size {distance_matrix.shape[0]} \"\n",
        "                         f\"does not match sequence names count {len(sequence_names)}\")\n",
        "\n",
        "    condensed_dist = squareform(distance_matrix)\n",
        "    Z = hierarchy.linkage(condensed_dist, method='average')\n",
        "\n",
        "    # Create and save the main dendrogram\n",
        "    plt.figure(figsize=(20, 5))\n",
        "    dendro = hierarchy.dendrogram(Z, labels=sequence_names, leaf_rotation=90, leaf_font_size=10)\n",
        "    plt.title(\"Sequence Similarity Tree\")\n",
        "    plt.xlabel(\"Sequence ID\")\n",
        "    plt.ylabel(\"Distance\")\n",
        "    plt.tight_layout()\n",
        "    tree_plot_path = os.path.join(output_dir, 'sequence_tree.png')\n",
        "    plt.savefig(tree_plot_path, dpi=300)\n",
        "    plt.close()\n",
        "    print(f\"Tree visualization saved to: {tree_plot_path}\")\n",
        "\n",
        "    # Define clusters at a certain level\n",
        "    cluster_labels = hierarchy.fcluster(Z, num_clusters, criterion='maxclust')\n",
        "\n",
        "    # Create individual cluster visualizations\n",
        "    for cluster_id in range(1, num_clusters + 1):\n",
        "        # Identify which sequences belong to this cluster\n",
        "        cluster_sequences = [sequence_names[i] for i in range(len(sequence_names)) if cluster_labels[i] == cluster_id]\n",
        "\n",
        "        # Plot the dendrogram for this cluster\n",
        "        plt.figure(figsize=(20, 5))\n",
        "        hierarchy.dendrogram(Z, labels=sequence_names, leaf_rotation=90, leaf_font_size=10,\n",
        "                             color_threshold=0, above_threshold_color='black', below_threshold_color='black')\n",
        "\n",
        "        # Highlight the cluster in a different color\n",
        "        for i, label in enumerate(sequence_names):\n",
        "            if label in cluster_sequences:\n",
        "                plt.gca().get_children()[i].set_color(f\"C{cluster_id}\")\n",
        "\n",
        "        plt.title(f\"Cluster {cluster_id}\")\n",
        "        plt.xlabel(\"Sequence ID\")\n",
        "        plt.ylabel(\"Distance\")\n",
        "        plt.tight_layout()\n",
        "\n",
        "        cluster_plot_path = os.path.join(output_dir, f'cluster_{cluster_id}.png')\n",
        "        plt.savefig(cluster_plot_path, dpi=300)\n",
        "        plt.close()\n",
        "        print(f\"Cluster {cluster_id} visualization saved to: {cluster_plot_path}\")\n",
        "\n",
        "def analyze_sequences(input_file, output_dir):\n",
        "    \"\"\"Run complete sequence analysis pipeline.\"\"\"\n",
        "    os.makedirs(output_dir, exist_ok=True)\n",
        "    extracted_file = os.path.join(output_dir, \"extracted_sequences.fasta\")\n",
        "    sequences = extract_designed_sequences(input_file, extracted_file)\n",
        "    sequence_list = [str(seq.seq) for seq in sequences]\n",
        "    sequence_names = [seq.id for seq in sequences]\n",
        "\n",
        "    print(\"Encoding sequences...\")\n",
        "    encoded_seqs = one_hot_encode(sequence_list)\n",
        "    print(\"Computing distance matrix...\")\n",
        "    distance_matrix = compute_distance_matrix(encoded_seqs)\n",
        "\n",
        "    print(\"Generating similarity heatmap...\")\n",
        "    plt.figure(figsize=(10, 10))\n",
        "    sns.heatmap(1 - distance_matrix[:50, :50], cmap=\"viridis\", square=True)\n",
        "    plt.title(\"Sequence Similarity Matrix\")\n",
        "    plt.tight_layout()\n",
        "    plt.savefig(os.path.join(output_dir, \"similarity_heatmap.png\"), dpi=300)\n",
        "    plt.close()\n",
        "\n",
        "    print(\"Creating phylogenetic tree...\")\n",
        "    create_phylogenetic_tree(distance_matrix, sequence_names, output_dir)\n",
        "\n",
        "    print(\"Analysis complete. Results saved in\", output_dir)\n",
        "    return sequences, distance_matrix\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    input_file = \"/content/drive/MyDrive/Fasta-files/3NOB_90-110/3NOB_90-110_design.fasta\"\n",
        "    output_dir = \"/content/drive/MyDrive/Fasta-files/3NOB_90-110/analysis_output\"\n",
        "    analyze_sequences(input_file, output_dir)\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "metadata": {
        "id": "HeSDPO33tOPF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "UzUBMUKRxWk9"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}